### PR TITLE
C++ API: torch::nn::BatchNorm1d

### PR DESCRIPTION
**Pull request by nuka137**
Add torch::nn::BatchNorm1d function/module support for the C++ API.
torch::nn::BatchNorm{2,3}d will be added after this PR is merged.

Related Issue: #25883 

Reviewer: @yf225 

I would like to discuss about below items.

* Necessity of `num_batches_tracked` in `BatchNormImplBase`
  * `num_batches_tracked` is needed to calculate `momentum` when we do not feed `momentum` argument in Python API. But in C++ API, `momentum` argument has a default value.
  * `num_batches_tracked` is only used for counting up `BatchNorm1d::foward()` call. I think it is no necessary for user anymore.
* The design of `BatchNorm{1,2,3}dOptions`
  * We have already `BatchNormOptions` used for deprecated `BatchNorm` module. However, it is hard to use it for `BatchNorm{1,2,3}dOptions` because of the arguments disagreement of each modules.
  * In this PR, I introduce `BatchNormOptionsv2` template class for the `BatchNorm{1,2,3}dOptions`. But I'm not sure this design is good or not.

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 1]
```diff
@@ -72,5 +78,75 @@ Tensor BatchNormImpl::pure_forward(
...
+         << "eps=" << options.eps() << ", "
+         << "momentum=" << options.momentum() << ", "
+         << "affine=" << options.affine() << ", "
+         << "track_running_stats=" << options.track_running_stats() << ")";
+}
+
+Tensor BatchNorm1dImpl::forward(const Tensor& input) {
+  TORCH_CHECK(
+      input.dim() != 2 && input.dim() !=3,
+      "expected 2D or 3D input (got %dD input)", input.dim());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/batchnorm.cpp]: I feel that we can add a virtual function `_check_input_dim` to the `BatchNormImplBase` class, override its implementation in `BatchNorm1dImpl`, and call this `_check_input_dim` function from `forward`, to match the Python version even better :D

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 2]
```diff
@@ -72,5 +78,75 @@ Tensor BatchNormImpl::pure_forward(
...
+BatchNormImplBase<D, Derived>::BatchNormImplBase(const BatchNormOptionsv2<D>& options_)
+    : options(options_) {
+  if (options.affine()) {
+    weight = this->register_parameter("weight", torch::ones({options.num_features()}));
+    bias = this->register_parameter("bias", torch::zeros({options.num_features()}));
+  }
+  if (options.track_running_stats()) {
+    running_mean = this->register_buffer("running_mean", torch::zeros({options.num_features()}));
+    running_var = this->register_buffer("running_var", torch::ones({options.num_features()}));
+    num_batches_tracked = this->register_buffer("num_batches_tracked", torch::tensor(0, torch::dtype(torch::kLong)));
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/batchnorm.cpp]: I think we should keep `num_batches_tracked` as `momentum` can still be null in the C++ version :D

**Review by yf225 — COMMENTED** [review]
@nuka137 My sincere apologies for the delay and thanks so much for the investigation and the awesome work! I left some comments regarding the design choices. The scope of work is large for `BatchNorm{1,2,3}d` and thanks so much for working on it. :D

File: torch/csrc/api/include/torch/nn/options/batchnorm.h [hunk 3]
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
...
+  /// Changing this parameter after construction __has no effect__.
+  TORCH_ARG(int64_t, num_features);
+
+  /// The epsilon value added for numerical stability.
+  /// Changing this parameter after construction __is effective__.
+  TORCH_ARG(double, eps) = 1e-5;
+
+  /// A momentum multiplier for the mean and variance.
+  /// Changing this parameter after construction __is effective__.
+  TORCH_ARG(double, momentum) = 0.1;
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/options/batchnorm.h]: Thanks. I fixed it.

File: torch/csrc/api/include/torch/nn/options/batchnorm.h [hunk 4]
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
   TORCH_ARG(double, momentum) = 0.1;
 };
 
+template <size_t D>
+struct BatchNormOptionsv2 {
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/options/batchnorm.h]: Good suggestion. I use `BatchNormBaseOptions` instead of `BatchNormOptionsv2 `.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 5]
```diff
@@ -10,10 +12,14 @@
...
 #include <vector>
 
+namespace F = torch::nn::functional;
+
 namespace torch {
 namespace nn {
 
 BatchNormImpl::BatchNormImpl(const BatchNormOptions& options_) : options(options_) {
+  LOG(WARNING) << "torch::nn::BatchNorm module is deprecated."
+               << "Use BatchNorm{1,2,3}d instead.";
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: I agree with that. Change to `TORCH_WARN`.

File: torch/csrc/api/include/torch/nn/modules/batchnorm.h [hunk 6]
```diff
@@ -75,5 +75,58 @@ class TORCH_API BatchNormImpl : public torch::nn::Cloneable<BatchNormImpl> {
...
+};
+
+/// Applies the BatchNorm1d function.
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.BatchNorm1d to learn
+/// about the exact behavior of this module.
+class TORCH_API BatchNorm1dImpl : public BatchNormImplBase<1, BatchNorm1dImpl> {
+ public:
+  using BatchNormImplBase<1, BatchNorm1dImpl>::BatchNormImplBase;
+
+  Tensor forward(const Tensor& input);
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/modules/batchnorm.h]: Thanks. I fixed it.
Also, I fixed the arguments of `F::batch_norm` too.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp (hunk 1 above)
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: OK. I added pure virtual function `_check_input_dim` to `BatchNormImplBase` class.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp (hunk 2 above)
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: Thanks. I added the codes if `options.track_running_stats()` is `false`.
Do these codes match your intention?

**Comment by nuka137** [comment]
@yf225 

Thanks for reviewing. I fixed all issues.
And I found that there is no `PrettyPrintBatchNorm1d` test in previous commit, so I added it in new commit.
Could you check it together?

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+#include <torch/cuda.h>
+
+namespace torch {
+namespace nn {
+namespace functional {
+
+inline Tensor batch_norm(const Tensor& input, const Tensor& running_mean,
+                         const Tensor& running_var, const Tensor& weight,
+                         const Tensor& bias, bool training,
+                         double momentum, double eps) {
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/batchnorm.h]: I believe the functional `batch_norm` can take some default values:
```python
def batch_norm(input, running_mean, running_var, weight=None, bias=None,
               training=False, momentum=0.1, eps=1e-5):
```
it would be awesome to do the same in C++ as well :D (we can use `Tensor()` as null value for `const Tensor&` type)

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  if (training) {
+    std::vector<int64_t> size = input.sizes().vec();
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/batchnorm.h]: I suspect we could do
```suggestion
    auto size = input.sizes();
```
under the hood it uses `ArrayRef` as type for `size` :D

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  return torch::batch_norm(
+    input,
+    weight,
+    bias,
+    running_mean,
+    running_var,
+    training,
+    momentum,
+    eps,
+    torch::cuda::cudnn_is_available());
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/batchnorm.h]: I think the C++ equivalent of `torch.backends.cudnn.enabled` is `at::globalContext().userEnabledCuDNN()`

File: torch/csrc/api/include/torch/nn/functional.h [hunk 10]
```diff
@@ -7,3 +7,4 @@
 #include <torch/nn/functional/normalization.h>
 #include <torch/nn/functional/pooling.h>
 #include <torch/nn/functional/vision.h>
+#include <torch/nn/functional/batchnorm.h>
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional.h]: It would be awesome to put it above `#include <torch/nn/functional/distance.h>`, to sort them alphabetically :D

File: test/cpp/api/modules.cpp [hunk 11]
```diff
@@ -965,6 +965,44 @@ TEST_F(ModulesTest, BatchNormPureForward) {
...
+TEST_F(ModulesTest, BatchNorm1dStateless) {
+  BatchNorm1d bn(BatchNorm1dOptions(5).track_running_stats(false).affine(false));
+
+  ASSERT_FALSE(bn->running_mean.defined());
+  ASSERT_FALSE(bn->running_var.defined());
+  ASSERT_FALSE(bn->num_batches_tracked.defined());
+  ASSERT_FALSE(bn->weight.defined());
+  ASSERT_FALSE(bn->bias.defined());
+}
+
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: We might need a forward+backward test for `BatchNorm1d` as well, similar to other modules such as `TEST_F(ModulesTest, MultiMarginLoss)`

File: test/cpp/api/functional.cpp [hunk 12]
```diff
@@ -892,3 +892,19 @@ TEST_F(FunctionalTest, Threshold) {
...
+
+  auto input = torch::randn({2, 5});
+  auto mean = torch::randn(5);
+  auto variance = torch::rand(5);
+  auto weight = torch::ones({num_features});
+  auto bias = torch::zeros({num_features});
+  auto output = F::batch_norm(input, mean, variance, weight, bias, false,
+                              momentum, eps);
+  auto expected = (input - mean) / torch::sqrt(variance + eps);
+  ASSERT_TRUE(output.allclose(expected));
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: It would be awesome to add a `F::batch_norm(input, mean, variant)` test for only the required arguments :D

**Review by yf225 — COMMENTED** [review]
@nuka137 Thanks so much and really appreciated the awesome work! I left some comments.

**Comment by yf225** [comment]
It would be awesome to update the `BatchNorm1d` entry in `test/cpp_api_parity/parity-tracker.md` as well. Thanks so much for the great work!

File: test/cpp/api/functional.cpp (hunk 12 above)
Comment (nuka137) [reply test/cpp/api/functional.cpp]: Added `BatchNormal1dDefaultOptions` test.

File: test/cpp/api/modules.cpp (hunk 11 above)
Comment (nuka137) [reply test/cpp/api/modules.cpp]: OK. I added `BatchNorm1d` test to do a forward/backward test.

File: torch/csrc/api/include/torch/nn/modules/batchnorm.h [hunk 13]
```diff
@@ -75,5 +75,64 @@ class TORCH_API BatchNormImpl : public torch::nn::Cloneable<BatchNormImpl> {
 /// module storage semantics.
 TORCH_MODULE(BatchNorm);
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ BatchNorm ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Base class for all (dimension-specialized) batchnorm modules.
+template <size_t D, typename Derived>
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/modules/batchnorm.h]: I think template parameter `D` is still needed to print a dimension in `pretty_print` method.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 14]
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
...
+  }
+  if (options.track_running_stats()) {
+    running_mean = this->register_buffer("running_mean", torch::zeros({options.num_features()}));
+    running_var = this->register_buffer("running_var", torch::ones({options.num_features()}));
+    num_batches_tracked = this->register_buffer("num_batches_tracked", torch::tensor(0, torch::dtype(torch::kLong)));
+  } else {
+    running_mean = this->register_buffer("running_mean", Tensor());
+    running_var = this->register_buffer("running_var", Tensor());
+    num_batches_tracked = this->register_buffer("num_batches_tracked", Tensor());
+  }
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: Moved all logic to `reset` method, and deleted `reset_parameters` method and `reset_track_running_stats` method.

File: torch/csrc/api/include/torch/nn/functional.h (hunk 10 above)
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/functional.h]: OK. I followed your instruction.

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h (hunk 9 above)
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/functional/batchnorm.h]: I replaced `torch.backends.cudnn.enabled` to `at::globalContext().userEnabledCuDNN()`.

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h (hunk 8 above)
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/functional/batchnorm.h]: Thanks. I followed your instruction.

File: torch/csrc/api/include/torch/nn/functional/batchnorm.h (hunk 7 above)
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/functional/batchnorm.h]: Thanks. I fixed it.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 15]
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
...
+  reset();
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::reset_parameters() {
+  reset_running_stats();
+  if (options.affine()) {
+    torch::nn::init::ones_(weight);
+    torch::nn::init::zeros_(bias);
+  }
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: Moved all logic to `reset` method, and deleted `reset_parameters` method and `reset_track_running_stats` method.

File: torch/csrc/api/src/nn/modules/batchnorm.cpp [hunk 16]
```diff
@@ -72,5 +78,102 @@ Tensor BatchNormImpl::pure_forward(
...
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::reset() {
+  reset_parameters();
+}
+
+template <size_t D, typename Derived>
+void BatchNormImplBase<D, Derived>::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::BatchNorm" << D << "d("
+         << "num_features=" << options.num_features() << ", "
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/src/nn/modules/batchnorm.cpp]: I fixed it.

File: torch/csrc/api/include/torch/nn/options/batchnorm.h [hunk 17]
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
   TORCH_ARG(double, momentum) = 0.1;
 };
 
+template <size_t D>
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/options/batchnorm.h]: I agree with that. I removed template parameter.

File: torch/csrc/api/include/torch/nn/options/batchnorm.h [hunk 18]
```diff
@@ -30,5 +30,34 @@ struct TORCH_API BatchNormOptions {
...
+  /// Changing this parameter after construction __has no effect__.
+  TORCH_ARG(bool, affine) = true;
+
+  /// Whether to store and update batch statistics (mean and variance) in the
+  /// module.
+  /// Changing this parameter after construction __has no effect__.
+  TORCH_ARG(bool, track_running_stats) = true;
+};
+
+using BatchNorm1dOptions = BatchNormBaseOptions<1>;
```
[the remark this answers is not in the document]
Comment (nuka137) [reply torch/csrc/api/include/torch/nn/options/batchnorm.h]: Thanks. I fixed it.

**Comment by nuka137** [comment]
@yf225 

Thanks for your instructions.
I followed them except for templatize `BatchNormImplBase`.
I think `BatchNormImplBase` is still needed to output dimension in `pretty_print`.
What do you think about it?

**Comment by nuka137** [comment]
@yf225 

Hi, I found that there is a change in `PrettyPrintHardtanh` test in this PR.
I believe that this is a accident from commit https://github.com/pytorch/pytorch/pull/28176/commits/ada8d540259ce7318331173100cd38dea019657f because `PrettyPrintHardtanh` test is failed now.
If you don't mind, I will fixed this in PR.

**Review by yf225 — APPROVED** [review]
@nuka137 Thanks so much for the awesome work! I took the liberty to merge `BatchNormBaseOptions` into `BatchNormOptions` because after thinking about it more I feel that forcing `F::batch_norm` to use `BatchNormBaseOptions` might not be a good idea, aesthetically speaking. Instead I renamed `features` to `num_features` and `stateful` to `track_running_stats` in `BatchNormOptions`, which is backward-compatibility breaking, but I think is better for the long term.
